Wake loss analysis method

ejsimley opened the pull request:
This PR contributes a wake loss analysis class `WakeLosses` in the analysis folder. The class is used to estimate wake losses during the period of record based on turbine SCADA data (and optionally met tower data) and estimate the long-term corrected wake losses using reanalysis data. Wake losses are estimated at the wind plant level as well as the individual turbine level and can be computed with or without uncertainty quantification. The class also contains methods for plotting the wake losses as a function of wind direction or wind speed.

In addition to the new wake losses class, the PR contains the following:

- An example notebook `06_wake_loss_analysis` demonstrates the wake loss analysis method using the LHB data set
- Added `asset_direction_matrix` and `get_freestream_turbines` functions to the `plant` module to determine free stream turbines for a particular wind direction to support the wake losses method
- Added the option to only return values for specific asset types in `asset_distance_matrix`
- a new threshold option (median absolute deviation) is added to the `bin_filter` function in the `filters` module

TODO: Still need to write some tests for the wake loss method, but wanted to start by submitting the rest of the code

ejsimley commented:
@RHammond2 FYI, I added some regression tests for the wake losses method

ejsimley commented:
Note that, at the risk of making this PR too long, I added some additional descriptions to the sphinx documentation to address some items in issue #214

RHammond2 reviewed (commented):
Thanks for getting this together, Eric! I think you did a really great job of bringing in the spirit of the v3 changes while making an entirely new method, which is definitely a daunting task, so kudos for that! Generally, this looks really good and there isn't much that I'd change at this time, though I've left some comments for areas that could be updated and have submitted a number of commits to clean up a few things, but nothing too much. With addressing a few of these changes I think that this could be ready to go with a pretty quick turn around unless you want to address some of the more substantive changes at this stage that are really pointed at all the analysis methods. Aside from some specific comments in the code, I've left a few general comments here as well and outlined the changes I made and why.

Tests that are failing:
- `test_wake_losses_with_UQ`
- `test_wake_losses_with_UQ_new_parameters`
- Ironically all the other regression tests pass, which is exceedingly rare for me.
- Not a test, but fits nicely here, the bibtex file seems to be missing so the documentation build creates errors for the unfound references, so if you could just add that in, then the problem should go away.

I went ahead and made the following changes (that are well separated by commits in case you disagree with any of them) to reduce some of the back and forth:
- minor typos or other formatting issues
- I did some more robust error checking on the `asset_type` input to ensure we can use helpful error messages
- I changed the rotor diameter normalization to ensure the order of the turbines is the same as the turbine distance matrix. In theory this should be the same by default, but in practice it's best to ensure these things are consistent so our math is guaranteed to be accurate.
- I chunked up your class docstring to be a numbered list like the other analysis methods and enforced a 100 character limit. It's probably just me still being sick, but I had a hard time reading it in paragraph form, so happy to revert the commit if you'd prefer the original format.
- In the docstrings throughout I did some sphinx formatting for attribute names and included roughly enforced the 100 character limit while I was in there, which also ensured that I read the docstring content to find typos and other small edits.
- `_num_sim` -> `num_sim` to match the other analysis classes
- Minor notebook reformatting to make the print statements shorter, catch returned items from plotting functions and ensure plots show up consistently in the notebook

I also have a few comments/suggestions for the notebook:
- I really like that you included the definitions of the different parameters/terms in the notebook to help explain everything.
- You have some really thorough explanations for what's going on in the plots and analyses which are greatly appreciated and a nice addition for introducing the method to users.
- Plotting routines included in the notebook should become plotting methods, likely in the `utils/plot.py` module.

ejsimley commented on openoa/plant.py:
```diff
@@ -71,6 +70,20 @@
+                "id",
+                "windspeed",
+                "power",
+            ],
```
Done

ejsimley commented on openoa/plant.py:
```diff
@@ -398,6 +411,23 @@ def rename_columns(df: pd.DataFrame, col_map: dict, reverse: bool = True) -> pd.
     return df.rename(columns=col_map)
 
 
+def wrap_180(x: float | np.ndarray):
```
Added a test for this function and also moved it to the `met_data_processing` utils module

ejsimley commented on openoa/plant.py:
```diff
@@ -1702,19 +1737,28 @@ def asset_ids(self) -> np.ndarray:
 
-    @cached_property
-    def asset_distance_matrix(self) -> pd.DataFrame:
+    def asset_distance_matrix(self, asset_type: str = None) -> pd.DataFrame:
```
As we discussed, I added `turbine_xx_matrix` and `tower_xx_matrix` methods that access data from the pre-calculated `asset_distance_matrix` and `asset_direction_matrix` attributes, so these don't get calculated every time the methods are called.

ejsimley commented on openoa/analysis/wake_losses.py:
```diff
@@ -0,0 +1,1584 @@
+                    )
+                )
+                % 360.0
+            )
```
Good suggestion. I moved this to `met_data_processing` and also added a test for it.

ejsimley commented on openoa/analysis/wake_losses.py:
```diff
@@ -0,0 +1,1584 @@
+            df_rean.index = df_rean.index.floor("H")
+
+            # Upsample to 10-minute samples to match SCADA data
+            df_rean = df_rean.resample("10T").ffill()
```
Good catch. I changed "10T" to `self.plant.metadata.scada.freq`

ejsimley commented on openoa/analysis/wake_losses.py:
```diff
@@ -0,0 +1,1584 @@
+            energy_lt_ws,
+        )
+
+    def plot_wake_losses_by_wind_direction(
```
I made the API similar to the other analysis class plotting functions by allowing customized inputs. I also moved the main plotting function to the `plot` utils module, similar to other analysis classes, to simplify the wake loss plotting functions in `wake_losses`.

RHammond2 replied:
Nice work on getting the two plotting routines here reduced to a single routine in the plot module so that each customized routine relies on a different data view going into the generalized routine!

ejsimley commented:
Hi @RHammond2, thanks again for the careful review and the improvements you made throughout. I think I have addressed your comments now with the latest updates. The tests are now all passing (I updated my python and package versions to match those used in the CI pipeline, which changed my local results slightly). I also added the missing bibtex file for the documentation.

Note that, as we discussed last week, I updated added `h5pyd` to the `setup.py` and `requirements.txt` files to fix the issue with the QA module mentioned in issue #220

RHammond2 commented:
@ejsimley I made a couple of really minor edits, so this is good to go whenever you're all set. I'll wait until Monday mid-morning to merge it in myself in case you have any other changes you think might be necessary/desired.

ejsimley commented:
Thanks @RHammond2! I will merge this in now.
